### PR TITLE
Single click pg_upgrade RFC

### PR DESCRIPTION
**Pull request by lubennikovaav**
RFC based on hackathon project

Needs review from product, storage and cplane teams to plan next steps

[rendered 039-single-click-pg-upgrade.md](
https://github.com/neondatabase/neon/blob/f62053b614364305edbc64eb740779dcaed9dd05/docs/rfcs/039-single-click-pg-upgrade.md)

See also slack channel `#proj-single-click-postgres-upgrade`

### CODE REVIEW

File: docs/rfcs/039-single-click-pg-upgrade.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+- How to restrict write access to the old branch after the siwtch?
+
+- How to communicate this pg_upgade switch to the user in UI / API?
+
+- How can we test the data correctness after the upgrade?
+
+- At what moment should we run recommended post-upgrade scripts?
+
+- How to handle extensions? This issue boils down to multi-version compute image.
```
Comment (koivunej) [inline docs/rfcs/039-single-click-pg-upgrade.md:199]: Interaction with `timeline_ancestor_detach`:

1. initially probably we want to just deny using this feature together with wal files from multiple postgres versions
    - or else we need to record the earlier pg version range which would be kind of extra
2. timeline ancestor detach in the presence of different pg versions will require rewriting all those pages which have not been modified in the branch instead of layer file copying
    - basically a reverse "image layer creation" operation we currently do with compaction
    - this might be already coming with archived branches or compaction enhancements

File: docs/rfcs/039-single-click-pg-upgrade.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+##### Where to set `S` for the new branch:
+
+- We need to use it as a start_lsn for the new Timeline struct
+- We need to set it in the control file of the new branch.
+- We need to set in in the Checkpoint record of the new branch.
+
+##### How to make the switch atomic:
+
+- We need to restrict writes to the old branch to get the `X` value and use it as `S` for the new branch.
+- IIUC, we need to do this right before the import. This means that timeline import time is customer visible pg_upgrade downtime. Can we somehow make it shorter? What is our target downtime for the upgrade? **Need storage team help here**
```
Comment (koivunej) [inline docs/rfcs/039-single-click-pg-upgrade.md:158]: I am unclear what help is needed here. Isn't the only problem that in our prototype we were thinking of only using the branch ancestor_lsn for selecting the pg version, not another lsn which would allow for these `pg_upgrade` Lsns in the created branch?

We also want to deny branching from these `pg_upgrade` Lsns, which in my above proposal would be `ancestor_lsn...pg_upgraded_lsn` and the `pg_upgraded_lsn` would be the first branchable (as in, to get a new branch with the latest version data). This `pg_upgraded_lsn` corresponds to your `S`, I think?

File: docs/rfcs/039-single-click-pg-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+1. OID and relfilenode:
+pg_upgrade ensures that all assignments of `pg_class.oid` (and `relfilenode`) so toast oids are the same between old and new clusters. 
+It also controls assignments of `pg_tablespace.oid`, `pg_type.oid`, `pg_enum.oid` and `pg_authid.oid`.
+
+To do this, it starts postgres in (undocumented) --binary-upgrade mode.
+
+2. Page layout of relation and SLRU
+
+Postgres aims to maintain backward compatibility of page layout and change it lazily in the new version.
+In exceptional cases, it provides a function to run during pg_upgrade the page layout.
```
Comment (hlinnaka) [inline docs/rfcs/039-single-click-pg-upgrade.md:82]:
```suggestion
In exceptional cases, it provides a function to run during pg_upgrade to rewrite pages to the new format.
```

We've never done that for relation files, but it has been done for some SLRUs in the past.  v17->v18 might involve some SLRU rewriting, thanks to the patches to move to 64-bit xids.

File: docs/rfcs/039-single-click-pg-upgrade.md (hunk 2 above)
Comment (hlinnaka) [reply docs/rfcs/039-single-click-pg-upgrade.md:158]: > Can we somehow make it shorter? What is our target downtime for the upgrade?

Let's make it work first, and try to optimize later. Downtime is OK for the first version, let's not get fixated on minimizing it.

File: docs/rfcs/039-single-click-pg-upgrade.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+##### How to make the switch atomic:
+
+- We need to restrict writes to the old branch to get the `X` value and use it as `S` for the new branch.
+- IIUC, we need to do this right before the import. This means that timeline import time is customer visible pg_upgrade downtime. Can we somehow make it shorter? What is our target downtime for the upgrade? **Need storage team help here** 
+
+
+## Known technical issues:
+
+- We mantain relsize cache in neon. We need to copy it over to the new cluster during upgrade.
```
Comment (hlinnaka) [inline docs/rfcs/039-single-click-pg-upgrade.md:163]: The relsize cache is just an in-memory construct. I don't think we need to do anything about it.

File: docs/rfcs/039-single-click-pg-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+- Fix import of the new version catalog into the new branch. We need to distinguish between catalog and non-catalog files and only skip the latter.
+Now this is done in a very hacky way, that will not work if any catalog files changed their relfilenode before the upgrade.
```
Comment (hlinnaka) [inline docs/rfcs/039-single-click-pg-upgrade.md:166]: pg_upgrade should have all the logic we need. It knows which relation files need to be copied/linked from the old cluster. That's the same list of relations that we need to preserve from the old branch, and everything else should be imported.

File: docs/rfcs/039-single-click-pg-upgrade.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+**Need compute (postgres) team help here**
+
+- We need to update XIDs in the new version controlfile and skip SLRU files during timeline import. See copy_xact_xlog_xid() for details.
```
Comment (hlinnaka) [inline docs/rfcs/039-single-click-pg-upgrade.md:169]: As first version, perhaps easiest to just import all SLRUs from the after-pg_upgrade cluster. Instead of trying to figure out which ones can be preserved.

File: docs/rfcs/039-single-click-pg-upgrade.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    - special mode of compute image (requires multi-version compute image).
+    - separate microservice running on pageserver machine. Synchronization with old running compute will be complicated here.
+
+## Open questions:
+
+**Need product, storage, compute and cplane teams help here**
+
+- Do we want to implement a full switch to the new version branch? Or do we want to start with "Test new version in a branch" feature? 
+Do we see product value in this feature?
+It is simpler to implenent, but we need to ensure that users won't use this new-version branch in production. How to do this?
```
Comment (hlinnaka) [inline docs/rfcs/039-single-click-pg-upgrade.md:187]: What's the harm if users use the new-version branch in production?

**Review by hlinnaka — APPROVED** [review]
